python_modules/dagster/dagster/core/execution/run_cancellation_thread.py

```diff
@@ -0,0 +1,36 @@
+from dagster import check
+import threading
+import time
+
+from dagster.core.instance import DagsterInstance
+from dagster.core.storage.pipeline_run import PipelineRunStatus
+from dagster.utils import send_interrupt
+
+INTERVAL_SECONDS = 10
+
+
+def _kill_on_cancel(instance: DagsterInstance, run_id):
+    check.inst_param(instance, "run_id", DagsterInstance)
+    check.str_param(run_id, "run_id")
+
+    while True:
+        time.sleep(INTERVAL_SECONDS)
+        status = instance.get_run_by_id(run_id).status
+        if status in [
+            PipelineRunStatus.CANCELING,
+            PipelineRunStatus.CANCELED,
+        ]:
+            send_interrupt()
+
+
+def start_run_cancellation_thread(instance, run_id):
+    int_thread = threading.Thread(
+        target=_kill_on_cancel,
+        args=(
+            instance,
+            run_id,
+        ),
+        name="kill-on-cancel",
+        daemon=True,
+    )
+    int_thread.start()
```

Comment:
> A thread can be flagged as a “daemon thread”. The significance of this flag is that the entire Python program exits when only daemon threads are left

This makes me think it's ok to not clean up

> Daemon threads are abruptly stopped at shutdown. Their resources (such as open files, database transactions, etc.) may not be released properly. If you want your threads to stop gracefully, make them non-daemonic and use a suitable signalling mechanism such as an Event.

This could be problematic especially if the thread has it's own instance

Reply:
yeah maybe event is the way to go, see e.g. the daemon shutdown path for an example of that